File: python/src/semantic_retrieval/generator/retrieval_augmented_generation/rag_completion_generator.py
from abc import abstractmethod
from typing import Generic, List

from attr import dataclass
from semantic_retrieval.access_control.access_passport import AccessPassport
from semantic_retrieval.common.types import P, Q, R
from semantic_retrieval.document.document import Document
from semantic_retrieval.generator.completion_generator import (
    LLMCompletionGenerator,
    LLMCompletionGeneratorParams,
)
from semantic_retrieval.prompts.prompt_template import PromptTemplate
from semantic_retrieval.retrieval.document_retriever import DocumentRetriever


@dataclass
class RAGCompletionGeneratorParams(
    Generic[P, Q], LLMCompletionGeneratorParams[P]
):
    retriever: DocumentRetriever[List[Document], Q]
    accessPassport: AccessPassport
    ragPromptTemplate: PromptTemplate


DEFAULT_RAG_TEMPLATE = "Answer the question based on the context below.\n\nContext:\n\n{{context}}\n\nQuestion: {{prompt}}\n\nAnswer:"


class T_(Generic[P, Q], RAGCompletionGeneratorParams[P, Q]):
    def __init__(self, a: P, b: Q):
        self.a = a
        self.b = b


class CompletionModelResponse:
    pass


class RetrieverQuery(Generic[R]):
    pass


class RAGCompletionGenerator(
    Generic[R, P],
    LLMCompletionGenerator[P, R],
):
    @abstractmethod
    async def get_retrieval_query(self, params: P) -> RetrieverQuery[R]:
        pass

    async def run(self, params: P) -> CompletionModelResponse:  # type: ignore
        raise NotImplementedError()

File: src/generator/retrieval-augmented-generation/ragCompletionGenerator.ts
import { AccessPassport } from "../../access-control/accessPassport";
import { PromptTemplate } from "../../prompts/prompt-templates/promptTemplate";
import { Document } from "../../document/document";
import { DocumentRetriever } from "../../retrieval/documentRetriever";
import {
  LLMCompletionGeneratorParams,
  LLMCompletionGenerator,
} from "../completionGenerator";

export interface RAGCompletionGeneratorParams<P, Q>
  extends LLMCompletionGeneratorParams<P> {
  retriever: DocumentRetriever<Document[], Q>;
  accessPassport?: AccessPassport;
  ragPromptTemplate?: PromptTemplate;
}

export const DEFAULT_RAG_TEMPLATE =
  "Answer the question based on the context below.\n\nContext:\n\n{{context}}\n\nQuestion: {{prompt}}\n\nAnswer:";

/**
 * A basic RAG completion generator that uses a retriever to retrieve documents which can
 * be leveraged for modifying the prompt prior to completion generation by the model
 */
export abstract class RAGCompletionGenerator<
  P,
  Q,
  R,
  T extends RAGCompletionGeneratorParams<P, Q>,
> extends LLMCompletionGenerator<P, R> {
  /**
   * Construct the query for the underlying retriever using the given parameters
   * @param params The parameters to use for constructing the query
   * @returns A promise that resolves to the query in valid format for the retriever
   */
  abstract getRetrievalQuery(params: T): Promise<Q>;

  /**
   * Performs completion generation using the given parameters and returns the generated
   * response data
   * @param params The parameters to use for generating the completion
   * @returns A promise that resolves to the generated completion data
   */
  async run(params: T): Promise<R> {
    const { accessPassport, prompt, retriever, ...modelParams } = params;

    const queryPrompt =
      typeof prompt === "string" ? prompt : await prompt.toString();

    const contextDocs = await retriever.retrieveData({
      accessPassport,
      query: await this.getRetrievalQuery(params),
    });

    const contextChunksPromises = [];
    for (const doc of contextDocs) {
      for (const fragment of doc.fragments) {
        contextChunksPromises.push(fragment.getContent());
      }
    }

    const context = (await Promise.all(contextChunksPromises)).join("\n");
    const ragPromptTemplate =
      params.ragPromptTemplate ?? new PromptTemplate(DEFAULT_RAG_TEMPLATE);

    ragPromptTemplate.setParameters({
      prompt: queryPrompt,
      context,
    });

    const response = await this.model.run({
      ...modelParams,
      prompt: ragPromptTemplate,
    });

    await this.callbackManager?.runCallbacks({
      name: "onRunCompletionGeneration",
      params,
      response,
    });

    return response;
  }
}
